### PR TITLE
feat: Reject non authenticated web API HTTP requests

### PR DESCRIPTION
**Pull request by lasrinnil**
## Description

This PR will add JwtToken auth for the public WebApp endpoints.

## References

* #1088

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Configuration/ChargeCommandAcceptedReceiverConfiguration.cs [hunk 1]
```diff
@@ -19,6 +19,7 @@
 using GreenEnergyHub.Charges.FunctionHost.Common;
 using GreenEnergyHub.Charges.Infrastructure.Core.MessagingExtensions;
 using GreenEnergyHub.Charges.Infrastructure.Core.MessagingExtensions.Registration;
+using GreenEnergyHub.Charges.Infrastructure.Core.Registration;
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Configuration/ChargeCommandAcceptedReceiverConfiguration.cs:22]: Are all these added usings required?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/integrationtest.local.settings.json.sample [hunk 2]
```diff
@@ -1,4 +1,6 @@
 ﻿{
   "AZURE_KEYVAULT_URL": "<Integration test key vault url>",
-  "AZURE_SECRETS_KEYVAULT_URL": "<Key vault with secrets url>"
+  "AZURE_SECRETS_KEYVAULT_URL": "<Key vault with secrets url>",
+  "B2C_TENANT_ID": "<Azure B2C tenant id>",
+  "BACKEND_SERVICE_APP_ID": "<Backend service application id>"
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/integrationtest.local.settings.json.sample:5]: I would prefer not to add this configuration requirement. Can't you just read them from the test code?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Configuration/ServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        /// <summary>
+        /// Adds registration of ActorMiddleware, ActorContext and ActorProvider.
+        /// </summary>
+        /// <param name="serviceCollection">ServiceCollection container</param>
+        public static void AddUserContext(this IServiceCollection serviceCollection)
+        {
+            serviceCollection.AddScoped<UserMiddleware>();
+            serviceCollection.AddScoped<IUserContext, UserContext>();
+            serviceCollection.AddScoped<IUserProvider, FakeUserProvider>();
+        }
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Configuration/ServiceCollectionExtensions.cs:55]: This looks like a step towards authorization. How about removing it to keep the PR focused? I assumse it's actually not used(?)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/FakeUserProvider.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Threading.Tasks;
+using Energinet.DataHub.Core.App.Common.Abstractions.Users;
+
+namespace GreenEnergyHub.Charges.WebApi
+{
+    public class FakeUserProvider : IUserProvider
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/FakeUserProvider.cs:21]: Remove?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Startup.cs [hunk 5]
```diff
@@ -44,9 +46,27 @@ public void ConfigureServices(IServiceCollection services)
...
+                    Type = SecuritySchemeType.Http,
+                    Scheme = "bearer",
+                    Reference = new OpenApiReference { Type = ReferenceType.SecurityScheme, Id = "Bearer", },
+                };
+
+                c.AddSecurityDefinition("Bearer", securitySchema);
+
+                var securityRequirement = new OpenApiSecurityRequirement { { securitySchema, new[] { "Bearer" } }, };
+
+                c.AddSecurityRequirement(securityRequirement);
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Startup.cs:64]: Haven't tried it but looks like a great thing to add 👍

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Startup.cs [hunk 6]
```diff
@@ -44,9 +46,27 @@ public void ConfigureServices(IServiceCollection services)
...
             });
 
             services.AddQueryApi(Configuration);
+            services.AddJwtTokenSecurity();
+            services.AddUserContext();
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Startup.cs:69]: Remove?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Startup.cs [hunk 7]
```diff
@@ -63,6 +83,11 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
 
             app.UseRouting();
 
+            app.UseMiddleware<JwtTokenMiddleware>();
+            app.UseMiddleware<UserMiddleware>();
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Startup.cs:87]: Remove?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Startup.cs [hunk 8]
```diff
@@ -63,6 +83,11 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
...
+
+            app.UseAuthentication();
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.WebApi/Startup.cs:89]: What does this actually do in contrast to the `JwtTokenMiddleware`? My guess is that you don't actually use this.

**Review by BjarkeMeier — APPROVED** [review]

**Comment by BjarkeMeier** [comment]
I suggest another PR title like e.g. "Reject non authenticated web API HTTP requests".

**Review by HenrikSommer — APPROVED** [review]

**Pull request merged**